### PR TITLE
Custom capabilities

### PR DESCRIPTION
**Pull request by matheus23**
Depends on #27 

This is the remaining work split out from that PR.

---

Capability checking is meant to be extensible with UCANs, so that every institution adopting UCANs can implement their own capabilities with their custom semantics for their use cases.

There's some common pieces to implementing capabilities and they can be abstracted here. The interface you need to implement is this:

```ts
export interface CapabilitySemantics<A> {
  /**
   * Try to parse a capability into a representation used for
   * delegation & returning in the `capabilities` call.
   * 
   * If the capability doesn't seem to match the format expected
   * for the capabilities with the semantics currently defined,
   * return `null`.
   */
  tryParsing(cap: Capability): A | null
  /**
   * This figures out whether a given `childCap` can be delegated from `parentCap`.
   * There are three possible results with three return types respectively:
   * - `A`: The delegation is possible and results in the rights returned.
   * - `null`: The capabilities from `parentCap` and `childCap` are unrelated and can't be compared nor delegated.
   * - `CapabilityEscalation<A>`: It's clear that `childCap` is meant to be delegated from `parentCap`, but there's a rights escalation.
   */
  tryDelegating(parentCap: A, childCap: A): A | null | CapabilityEscalation<A>
  // TODO builders
}
```

For reference, `Capabilitiy` is just `{ [rsc: string]: unknown, cap: string }` and `CapabilityEscalation` is just a representation for escalation errors with a string message and a reference to the escalating capability.

These semantics can then be plugged into the `capabilities` function in `attenuation.ts`. This is its signature:
```ts
export function capabilities<A>(
  ucan: Chained,
  capability: CapabilitySemantics<A>,
): Iterable<CapabilityResult<A>>

export type CapabilityResult<A>
  = CapabilityWithInfo<A>
  | CapabilityEscalation<A>
```

`capabilities` returns an `Iterator`, so it's totally possible to short-circuit the capability parsing for any reason, e.g. because you've already verified that the UCAN gives the needed semantics.

You need to pass it a `Chained` value, i.e. a UCAN that has been parsed deeply using `Chained.fromToken`.

Each capability comes with some additional information attached to it, independent of the semantics used:
- An `originator` DID, i.e. which DID brought the capability "into life", so which DID claims to be the owner of said capability. This should then be checked by the service that checks the UCAN.
- An expiry time, i.e. the minimum of all expiry times from the UCANs involved in the delegation chain.
- A "not before" time, i.e. the maximum "not before" time from all the UCANs involved in the delegation chain.

In production you should filter out `CapabilityEscalations` from the `CapabilityResult`s you find during iterating the capabilities. But in development this helps debugging why your UCANs might not provide the capabilities you expect.

---

There's 3 semantics implemented right now:
1. A really simple email sending capability. I.e. you either have it or you don't.
2. The wnfs public capability. The semantics allow delegating access to only subdirectories or delegating reduced rights (e.g. only creation, but not deletion rights).
3. The wnfs private capability. This semantic is more involved, because you need to accumulate information about the delegation chain. You don't know in which way the `inumbers` (identification numbers for private directories/files) relate to each other (i.e. which is subdirectory of what).

The semantics for (1) is only useful for demonstration & testing (it *probably* shouldn't be used as-is for production), so it's in the `tests` directory.
The semantics for (2) and (3) will be used in production. However, should they really be part of ts-ucan? I'm not sure. For now, I've put them in there, but they're not exposed in the top-level API. Let me know what you think!

### CODE REVIEW

**Comment by matheus23** [comment]
Fixes #21 

This PR doesn't implement something like `compareAny` yet. The only (theoretical) use case for this is the "can opener" analogy, where you need both the "can" capability as well as the "opener" capability to open the can.

I'd love to have a real use for this before implementing that, so I'm putting that on hold for now. I can see that abstracted out in some way later, and have the current abstractions use the new one then.

File: src/capability/wnfs.ts [hunk 1]
```diff
@@ -0,0 +1,162 @@
+import { Capability } from "../types"
+import { capabilities, CapabilityEscalation, CapabilitySemantics } from "../attenuation"
+import { Chained } from "../chained"
+
+
+export const wnfsCapLevels = {
+  "SUPER_USER": 0,
+  "OVERWRITE": -1,
+  "SOFT_DELETE": -2,
```
Comment (dholms) [inline src/capability/wnfs.ts:9]: why negative numbers here?

**Review by dholms — APPROVED** [review]
Nice this looks great. And seeing it in action in the tests makes it really click. Really like the design around passing back the `info` along with the requested capability as well as passing both a `tryParse` and `tryDelegate` function.

I think it's fine to have the wnfs capabilities in this lib. It's the main use case for ucans right now, and it gives an example (that's a bit more in depth than email) of how to write the semantics functions right here in the codebase :ok_hand:

File: src/capability/wnfs.ts (hunk 1 above)
Comment (matheus23) [reply src/capability/wnfs.ts:9]: Really doesn't matter that much. But I thought that `SUPER_USER` is the ceiling of permissions, so it's the anchor point, thus 0. I.e. I don't expect anything to ever be on level 1.
At first I then made the other permission levels count up, but then the comparison checks are reverse. I.e. I'd expect `wnfsCapLevel[capBig] > wnfsCapLevel[capSmall]` to mean that `capBig` can do more than `capSmall`.
So, I just made them negative :man_shrugging:  :smile:

**Comment by matheus23** [comment]
> I think it's fine to have the wnfs capabilities in this lib. It's the main use case for ucans right now, and it gives an example (that's a bit more in depth than email) of how to write the semantics functions right here in the codebase :ok_hand:

It's actually possible I'll kick the private WNFS capabilities out, because it might actually be nicer to replace the `requiredINumbers: Set<string>` with `bareNameFilter: Uint8Array`. But then I'd need to pull in our wnfs2 bloom filters :confused: 

And then it would maybe be weird to have only the public wnfs capabilities in there, but not the private ones :thinking: I think ideally they're actually not in here, but instead there'd be more examples in the tests. E.g. just a simpler wnfs-unspecific path-based capability and maybe a capability for REST with URLs and HTTP methods as `cap`.

I'll leave them in for this PR though.

Felt cute, might delete later lol

**Pull request merged**